### ISSUE TITLE
XHR request redirected instead of 401

### ISSUE THREAD

**Issue by ace130-github** [oauth2-proxy/oauth2-proxy#1522]
## Expected Behavior

As per description of the [Behaviour](https://oauth2-proxy.github.io/oauth2-proxy/docs/behaviour) any request with header `Accept: application/json`  should **not** be redirected, but instead answered with 401 by oauth2-proxy.

## Current Behavior

I observe requests with the said header that are still redirected instead of 401'ed. Here's one such example:
![grafik](https://user-images.githubusercontent.com/14802412/150838940-6fe41258-91be-4643-b03b-e468da4552e3.png)

## Possible Solution

All requests with `Accept: application/json` should be answered with 401 if no proper auth is present.

## Steps to Reproduce (for bugs)

This happens sporadically, but appears to be associated with session inactivity timeout.
We're using the following settings:
* OAUTH2_PROXY_COOKIE_REFRESH=1m
* Session inactivity in Authserver (Keycloak): 2h
and some more. 

* OAUTH2_PROXY_PROVIDER=oidc

But this should not matter, as there are only two possibilities: Either the proxy recognizes a valid auth, then it should just proxy the request upstream, or it does not, in which case it should answer with 401. In any case, a redirect of a XHR request is unexpected and not useful.

However, if it helps in analysis, I can of course provide more information on our configuration.

## Your Environment

I suppose we have a pretty standard setup:
* A single page webapp, static resources (except index.html) are not protected (by `--skip-auth-regex`).
* API route and index.html are protected by the proxy. Hitting index.html without auth should trigger redirect to login (served by keycloak)
* After successful auth the webapp is doing mostly XHR requests to the api endpoint
* If session timeout occurs, the next XHR request should receive 401
* The webapp handles this 401 by reloading index.html, which again triggers login 

If a XHR request is instead answered with a 30x redirect, the webapp tries to follow this redirection, causing CORS issues and -even if the CORS issues were sorted out- is not helpful because the login page is not a valid response for the redirected api call.

## Version used:

7.2.1 in Docker

**Comment by ace130-github** [comment]
One speculation: 
Could this situation arise when the upstream server sends a 401 response?

In our setup the "cookie refresh time" is set to the same value as the access token lifetime (1m), and the access token is being sent upstream as per `--pass-access-token`.

Hypothesis:
* Browser sends XHR Request (Accept: application/json) with oauth2-proxy session cookie
* Proxy finds access token in session, which is barely before its expiry time. Since the cookie refresh time has the same value, there's also no reason to refresh it.
* Proxy sends request including access token upstream
* upstream API server checks token and finds it expired, thus responding with 401
* _oauth2-proxy now responds with a redirect to login URL_

This last step is pure speculation, but if this is possible, it would explain the "illegal" redirect of an XHR request in this issue report.

Question: Would oauth2-proxy do such a thing in the situation described? I have not found any documentation on how oauth2-proxy generally handles 401s from upstream.

**Comment by SeWieland** [comment]
I do have the exact same issue and it _does_ lead to some nasty bugs.
And no, it happens every time, each request, even if `--force-json-errors is set` or the header is added manually.

I suspect the issue here that I'm using an nginx ingress with annotiations such as: 
```
nginx.ingress.kubernetes.io/auth-signin: "https://$host/oauth2/sign_in"
nginx.ingress.kubernetes.io/auth-url: "https://$host/oauth2/auth"
```

As far as I understood, nginx checks against the auth-url if the request is authenticated and if not - sends a 302 to the signin page with a proper redirect to the original resource...

So if there is a reverse-proxy in front of oauth2-proxy, this mechanism cannot work.

However I'm very unsure if my understanding is corret and I'm further unaware of a solution for this problem :(

**Comment by SeWieland** [comment]
Update:

So yes. removing the ingress annotation `nginx.ingress.kubernetes.io/auth-signin` completely for all "backend" resources works for me quite well with nginx-ingress.

Only the user facing services / frontends need the signin redirection 👍 

This split in "behaviour" by ingress routes is fine enough for me.
The AJAX detection by headers is not really needed in this way.

**Comment by ppawiggers** [comment]
I experience the same issue. I send a `curl` request without cookies to the protected backend (with the `accept` header), but it's still 302 redirect to the sign in page.

I don't want to remove the `auth-signin` annotation on the backend, because that also hosts a Swagger UI for developers to use, and they should be redirected to the sign in page.

**Comment by braunsonm** [comment]
Unless I'm missing something this seems to be an issue with your architecture or configuration. I don't see any specific reproduction instructions. If you provide a reproduction repo we might be able to look at if this is a bug in OAuth proxy.

Reopen if you do believe this is a bug with oauth proxy

**Comment by geudrik** [comment]
Yeah, this issue shouldn't have been closed, I agree. Returning a 302 for CORS requests is .. bad. It should be a 401

**Comment by zoriax** [comment]
Hi everyone,

It's exactly the same issue as mentioned in my "old" post #2072 

I agree, a 401 should be THE solution ;-) and as @geudrik said it should be "easy" to correct and implement.

**Comment by JoelSpeed** [comment]
Has anyone looked at raising a PR to resolve this, I understand it's an on-going issue that's been around for a while now
